velox/functions/lib/Re2Functions.cpp

```diff
@@ -970,51 +969,143 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
+      isPreviousEscapeChar = false;
+    } else {
+      if (current == escapeChar) {
+        isPreviousEscapeChar = true;
+        continue;
+      } else {
+        str << current;
+      }
+    }
+  }
+
+  return str.str();
+}
+
+std::string unescape(StringView pattern, std::optional<char> escapeChar) {
+  return unescape(pattern, 0, pattern.size(), escapeChar);
+}
+
+bool PatternStringIterator::hasNext() {
+  return currentIndex_ < (int32_t)(pattern_.size() - 1);
+}
+
+void PatternStringIterator::next() {
+  currentIndex_++;
+  previousState_ = state_;
+
+  auto currentChar = pattern_.data()[currentIndex_];
+  state_.isEscaping = !previousState_.isEscaping && currentChar == escapeChar_;
+  state_.isWildcard = !previousState_.isEscaping &&
+      (currentChar == '_' || currentChar == '%') && currentChar != escapeChar_;
+}
+
+PatternMetadata determinePatternKind(
+    StringView pattern,
+    std::optional<char> escapeChar) {
   vector_size_t patternLength = pattern.size();
-  vector_size_t i = 0;
-  // Index of the first % or _ character.
+
+  // Index of the first % or _ character(not escaped).
   vector_size_t wildcardStart = -1;
   // Count of wildcard character sequences in pattern.
   vector_size_t numWildcardSequences = 0;
   // Index of the first character that is not % and not _.
   vector_size_t fixedPatternStart = -1;
   // Index of the last character in the fixed pattern, used to retrieve the
   // fixed string for patterns of type kSubstring.
-  vector_size_t fixedPatternEnd = 0;
+  vector_size_t fixedPatternEnd = -1;
   // Total number of % characters.
   vector_size_t anyCharacterWildcardCount = 0;
   // Total number of _ characters.
   vector_size_t singleCharacterWildcardCount = 0;
-  auto patternStr = pattern.data();
 
-  while (i < patternLength) {
-    if (patternStr[i] == '%' || patternStr[i] == '_') {
-      if (wildcardStart == -1) {
-        wildcardStart = i;
+  PatternStringIterator iterator{pattern, escapeChar};
+  bool fallbackToGeneric = false;
+
+  // Return false if failed to handle the new fixed pattern(need to fallback),
+  // true otherwise.
+  auto handleNewFixedPattern = [&]() -> bool {
+    // Record the first fixed pattern start.
+    if (fixedPatternStart == -1) {
+      fixedPatternStart = iterator.currentIndex();
+    } else {
+      // This is not the first fixed pattern, not supported, so fallback.
+      if (iterator.previousState().isWildcard) {
+        fallbackToGeneric = true;
+        return false;
       }
-      numWildcardSequences++;
-      // Look till the last contiguous wildcard character, starting from this
-      // index, is found, or the end of pattern is reached.
-      while (i < patternLength &&
-             (patternStr[i] == '%' || patternStr[i] == '_')) {
-        singleCharacterWildcardCount += (patternStr[i] == '_');
-        anyCharacterWildcardCount += (patternStr[i] == '%');
-        i++;
+    }
+
+    return true;
+  };
+
+  while (iterator.hasNext()) {
```

Comment:
I'm not following this logic. Would you explain what it does?

Reply:
Function `determinePatternKind` has two parts:

1. The first part collect stats about the pattern: e.g. where the first wildcard occurs, where the first fixedPattern occurs etc.
2. The second part deduce the optimized pattern based on the stats collected in the first part.

The change I made is mainly to the first part, I  introduced a dedicated struct: `PatternStringIterator`, the reason is : in the previous code, we advance the `cursor` by simply `i++`, now to support escape char we have some state to update when advancing the cursor:

```
  /// Advance the cursor to next char.
  void next() {
    currentIndex_++;
    previousCharKind_ = charKind_;

    auto currentChar = pattern_.data()[currentIndex_];
    if (previousCharKind_ != CharKind::kEscape && currentChar == escapeChar_) {
      charKind_ = CharKind::kEscape;
    } else if (
        previousCharKind_ != CharKind::kEscape && currentChar != escapeChar_ &&
        (currentChar == '_' || currentChar == '%')) {
      charKind_ = CharKind::kWildcard;
    } else {
      charKind_ = CharKind::kNormal;
    }
  }
```

Introducing `PatternStringIterator` will make the logic easier to read & maintain.